features/flows/elastic/src/main/java/org/opennms/netmgt/flows/elastic/thresholding/FlowThresholding.java

```diff
@@ -0,0 +1,135 @@
+    private final CollectionAgentFactory collectionAgentFactory;
+
+    private final Map<ApplicationKey, AtomicLong> applicationAccumulator;
+
+    private final Cache<NodeInterfaceKey, ThresholdingSession> thresholdingSessions;
+
+    public FlowThresholding(final ThresholdingService thresholdingService,
+                            final CollectionAgentFactory collectionAgentFactory,
+                            final CacheConfig sessionCacheConfig) {
+        this.thresholdingService = Objects.requireNonNull(thresholdingService);
+        this.collectionAgentFactory = Objects.requireNonNull(collectionAgentFactory);
+
+        //noinspection unchecked
+        this.applicationAccumulator = Maps.newConcurrentMap();
+
+        //noinspection unchecked
+        this.thresholdingSessions = new CacheBuilder<NodeInterfaceKey, ThresholdingSession>()
+                .withConfig(sessionCacheConfig)
+                .withCacheLoader(new CacheLoader<NodeInterfaceKey, ThresholdingSession>() {
+                    @Override
+                    public ThresholdingSession load(final NodeInterfaceKey nik) throws Exception {
+                        return FlowThresholding.this.thresholdingService.createSession(nik.nodeId,
+                                                                                       nik.ifaceAddr,
+                                                                                       SERVICE_NAME,
+                                                                                       FLOW_APP_RRD_REPO,
+                                                                                       new ServiceParameters(Collections.emptyMap()));
+                    }
+                }).build();
+    }
+
+    public void threshold(final List<FlowDocument> documents,
+                          final FlowSource source) throws ExecutionException, ThresholdInitializationException {
+        for (final var document : documents) {
+            if (document.getNodeExporter() != null && !Strings.isNullOrEmpty(document.getApplication())) {
+                final var nodeId = document.getNodeExporter().getNodeId();
+
+                final var key = new ApplicationKey(nodeId,
+                                                   document.getDirection() == Direction.INGRESS
+                                                   ? document.getInputSnmp()
+                                                   : document.getOutputSnmp(),
+                                                   document.getApplication());
+
+                // Update the counter
+                final var counter = this.applicationAccumulator.computeIfAbsent(key, k -> new AtomicLong(0))
+                                                               .addAndGet(document.getBytes());
+
+                // Apply thresholding
+                final var thresholdingSession = this.thresholdingSessions.get(new NodeInterfaceKey(nodeId, source.getSourceAddress()));
```

Comment:
Can you document why the `NodeInterfaceKey` uses the `sourceAddress` and not the interface index? Can the same `nodeId` go together with different `sourceAddress` values? Maybe this would be a good chance to document what a `FlowSource` is.

Reply:
> Can you document why the `NodeInterfaceKey` uses the `sourceAddress` and not the interface index?

Simply, because that's what the `ThresholdingSession` and the `CollectionAgent` need for creation.

> Can the same `nodeId` go together with different `sourceAddress` values?

Indeed. We use the source address (and location) to reverse-lookup the node which has this interface assigned.
This takes into account that the same interface may be assigned to multiple nodes (even in the same location) and that proxies may be used to aggregate multiple exporters into a single source (hence the support for `sourceId` lookup).

This is far from ideal. In case of `sourceId` lookups, we maybe choose an interface, that maybe does not exist on the interface. I've created https://issues.opennms.org/browse/NMS-13771 to track this later.

> Maybe this would be a good chance to document what a `FlowSource` is.

Not sure if the `FlowSource` needs documentation. It's a source of flows. The address (in a location) that sends out flows.

Reply:
Can the same flow source send out flows of different nodes? Your remark on "proxies" hints for that. In that case the `sourceAddress` seems of dubious use to identify a metric for thresholding.

The relation between `nodeId`, `sourceAddress`, `interfaces`, and `itfIndex` is still not clear to me. If `sourceAddress` and `location` are used to lookup a `nodeId` then having both, `nodeId` and `sourceAddress` as part of the `NodeInterfaceKey` seems unnecessary. Shouldn't the `location` be considered? I would like to have these relationships be documented somewhere.

I was surprised to hear that "the same interface may be assigned to multiple nodes". My current understanding is that "interface" and IP address (numerical) are often used as synonyms. Is this the case here, too? In my understanding there is a 1:1 relationship between (nodeId, itfIndex) and numeric IP addresses.

Can you point me to some documentation where the involved terms are documented? If there is no such place then maybe we should create a documentation issue.

Reply:
> Can the same flow source send out flows of different nodes? Your remark on "proxies" hints for that. In that case the `sourceAddress` seems of dubious use to identify a metric for thresholding.

Yes it can. Therefore we use the `nodeId` as part of the key. The `nodeId` is found using the `sourceId` and meta-data associations.

> The relation between `nodeId`, `sourceAddress`, `interfaces`, and `itfIndex` is still not clear to me. If `sourceAddress` and `location` are used to lookup a `nodeId` then having both, `nodeId` and `sourceAddress` as part of the `NodeInterfaceKey` seems unnecessary. Shouldn't the `location` be considered? I would like to have these relationships be documented somewhere.

There are two types of interfaces: IP Interfaces (identified by its IP) and SNMP Interfaces (identified by the SNMP interface index). A Node can have multiple IP Interfaces and multiple SNMP Interfaces. Two IP Interfaces with the same IP address can exists if they are assigned to different nodes, even if the nodes are in the same location.

As mentioned above, the proxy case is not implemented correctly right now. The perfect solution would be to used the database ID of the IP Interface. But lookup of this is expensive and requires larger restructuring.

The key was chosen this way to make allow creation of the cached value from the key only. Other solution would increase complexity in caching. I don't see a big problem here. The key may not be as small as it could be - but there are no duplicates possible.

This all is ages old core concept of OpenNMS. These internals are not documented AFAIK. But creating documentation is out-of-scope for this task.
 
> I was surprised to hear that "the same interface may be assigned to multiple nodes". My current understanding is that "interface" and IP address (numerical) are often used as synonyms. Is this the case here, too? In my understanding there is a 1:1 relationship between (nodeId, itfIndex) and numeric IP addresses.

There can be duplicates. Even in the same location.
In addition, "Interface" can mean two things: The IP interface on which we exchange things with and SNMP interfaces on witch we collect metrics.

> Can you point me to some documentation where the involved terms are documented? If there is no such place then maybe we should create a documentation issue.

Sorry, I can not. I don't think, this is part of the task. If you fell this needs documentation - go ahead and create such an issue.